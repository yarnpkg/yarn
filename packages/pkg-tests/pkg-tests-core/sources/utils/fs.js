/* @flow */

import type {Gzip} from 'zlib';

const fs = require('fs-extra');
const klaw = require('klaw');
const path = require('path');
const tarFs = require('tar-fs');
const tmp = require('tmp');
const zlib = require('zlib');

const miscUtils = require('./misc');

exports.walk = function walk(
  source: string,
  {filter, relative = false}: {|filter?: Array<string>, relative?: boolean|} = {},
): Promise<Array<string>> {
  return new Promise((resolve, reject) => {
    const paths = [];

    const walker = klaw(source, {
      filter: itemPath => {
        if (!filter) {
          return true;
        }

        const stat = fs.statSync(itemPath);

        if (stat.isDirectory()) {
          return true;
        }

        const relativePath = path.relative(source, itemPath);

        if (miscUtils.filePatternMatch(relativePath, filter)) {
          return true;
        }

        return false;
      },
    });

    walker.on('data', ({path: itemPath}) => {
      const relativePath = path.relative(source, itemPath);

      if (!filter || miscUtils.filePatternMatch(relativePath, filter)) {
        paths.push(relative ? relativePath : itemPath);
      }

      // This item has been accepted only because it's a directory; it doesn't match the filter
      return;
    });

    walker.on('end', () => {
      resolve(paths);
    });
  });
};

exports.packToStream = function packToStream(
  source: string,
  // $FlowFixMe Flow doesn't support null default parameters in exact types
  {virtualPath = null}: {|virtualPath?: ?string|} = {},
): Gzip {
  if (virtualPath) {
    if (!path.isAbsolute(virtualPath)) {
      throw new Error('The virtual path has to be an absolute path');
    } else {
      virtualPath = path.resolve(virtualPath);
    }
  }

  const zipperStream = zlib.createGzip();

  const packStream = tarFs.pack(source, {
    map: header => {
      if (true) {
        header.name = path.resolve('/', header.name);
        header.name = path.relative('/', header.name);
      }

      if (virtualPath) {
        header.name = path.resolve('/', virtualPath, header.name);
        header.name = path.relative('/', header.name);
      }

      return header;
    },
  });

  packStream.pipe(zipperStream);

  packStream.on('error', error => {
    zipperStream.emit('error', error);
  });

  return zipperStream;
};

exports.packToFile = function packToFile(target: string, source: string, options: *): Promise<void> {
  const tarballStream = fs.createWriteStream(target);

  const packStream = exports.packToStream(source, options);
  packStream.pipe(tarballStream);

  return new Promise((resolve, reject) => {
    tarballStream.on('error', error => {
      reject(error);
    });

    packStream.on('error', error => {
      reject(error);
    });

    tarballStream.on('close', () => {
      resolve();
    });
  });
};

exports.createTemporaryFolder = function createTemporaryFolder(): Promise<string> {
  return new Promise((resolve, reject) => {
    tmp.dir({unsafeCleanup: true}, (error, dirPath) => {
      if (error) {
        reject(error);
      } else {
        resolve(dirPath);
      }
    });
  });
};

exports.createTemporaryFile = async function createTemporaryFile(filePath: string): Promise<string> {
  if (filePath) {
    if (path.normalize(filePath).match(/^(\.\.)?\//)) {
      throw new Error('A temporary file path must be a forward path');
    }

    const folderPath = await exports.createTemporaryFolder();
    return path.resolve(folderPath, filePath);
  } else {
    return new Promise((resolve, reject) => {
      tmp.file({discardDescriptor: true}, (error, filePath) => {
        if (error) {
          reject(error);
        } else {
          resolve(filePath);
        }
      });
    });
  }
};

exports.writeFile = async function writeFile(target: string, body: string | Buffer): Promise<void> {
  await fs.mkdirp(path.dirname(target));
  await fs.writeFile(target, body);
};

exports.readFile = function readFile(source: string, encoding: ?string = null): Promise<any> {
  return fs.readFile(source, encoding);
};

exports.writeJson = function writeJson(target: string, object: any): Promise<void> {
  return exports.writeFile(target, JSON.stringify(object));
};

exports.readJson = async function readJson(source: string): Promise<any> {
  const fileContent = await exports.readFile(source, 'utf8');

  try {
    return JSON.parse(fileContent);
  } catch (error) {
    throw new Error(`Invalid json file (${source})`);
  }
};
