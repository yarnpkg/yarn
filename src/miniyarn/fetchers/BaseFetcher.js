export class BaseFetcher {
  supports(packageLocator, {env}) {
    // Return true if the resolver does support the package

    throw new Error(`Unimplemented supports strategy`);
  }

  async fetch(packageLocator, {env}) {
    // Fetch all data for a specified locator, then return an object { packageInfo, handler }
    //
    // - packageInfo is the full packageInfo referenced by the specified locator
    // - handler is a fsUtils.Handler instance that contains the path to something on the filesystem.

    throw new Error(`Unimplemented fetch strategy`);
  }
}
