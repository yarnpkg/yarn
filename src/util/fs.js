/* @flow */

import BlockingQueue from './blocking-queue.js';
import * as promise from './promise.js';
import {promisify} from './promise.js';
import map from './map.js';

const path = require('path');
const fs = require('fs');

export const lockQueue = new BlockingQueue('fs lock');

export const readFileBuffer = promisify(fs.readFile);
export const writeFile = promisify(fs.writeFile);
export const readlink = promisify(fs.readlink);
export const realpath = promisify(fs.realpath);
export const readdir = promisify(fs.readdir);
export const rename = promisify(fs.rename);
export const access = promisify(fs.access);
export const stat = promisify(fs.stat);
export const unlink = promisify(require('rimraf'));
export const mkdirp = promisify(require('mkdirp'));
export const exists = promisify(fs.exists, true);
export const lstat = promisify(fs.lstat);
export const chmod = promisify(fs.chmod);

const fsSymlink = promisify(fs.symlink);
const invariant = require('invariant');
const stripBOM = require('strip-bom');

const noop = () => {};

export type CopyQueueItem = {
  src: string,
  dest: string,
  onFresh?: ?() => void,
  onDone?: ?() => void,
};

type CopyQueue = Array<CopyQueueItem>;

type CopyFileAction = {
  type: 'file',
  src: string,
  dest: string,
  atime: number,
  mtime: number,
  mode: number
};

type CopySymlinkAction = {
  type: 'symlink',
  dest: string,
  linkname: string,
};

type CopyActions = Array<CopyFileAction | CopySymlinkAction>;

type PossibleExtraneous = void | false | Iterable<string>;

type CopyOptions = {
  onProgress: (dest: string) => void,
  onStart: (num: number) => void,
  possibleExtraneous: PossibleExtraneous,
  ignoreBasenames: Array<string>,
};

async function buildActionsForCopy(
  queue: CopyQueue,
  events: CopyOptions,
  possibleExtraneousSeed: PossibleExtraneous,
): Promise<CopyActions> {
  const possibleExtraneous: Set<string> = new Set(possibleExtraneousSeed || []);
  const noExtraneous = possibleExtraneousSeed === false;
  const files: Set<string> = new Set();

  // initialise events
  for (const item of queue) {
    const onDone = item.onDone;
    item.onDone = () => {
      events.onProgress(item.dest);
      if (onDone) {
        onDone();
      }
    };
  }
  events.onStart(queue.length);

  // start building actions
  const actions: CopyActions = [];

  // custom concurrency logic as we're always executing stacks of 4 queue items
  // at a time due to the requirement to push items onto the queue
  while (queue.length) {
    const items = queue.splice(0, 4);
    await Promise.all(items.map(build));
  }

  // remove all extraneous files that weren't in the tree
  if (!noExtraneous) {
    for (const loc of possibleExtraneous) {
      if (!files.has(loc)) {
        await unlink(loc);
      }
    }
  }

  return actions;

  //
  async function build(data): Promise<void> {
    const {src, dest} = data;
    const onFresh = data.onFresh || noop;
    const onDone = data.onDone || noop;
    files.add(dest);

    if (events.ignoreBasenames.indexOf(path.basename(src)) >= 0) {
      // ignored file
      return;
    }

    const srcStat = await lstat(src);
    let srcFiles;

    if (srcStat.isDirectory()) {
      srcFiles = await readdir(src);
    }

    if (await exists(dest)) {
      const destStat = await lstat(dest);

      const bothSymlinks = srcStat.isSymbolicLink() && destStat.isSymbolicLink();
      const bothFolders = srcStat.isDirectory() && destStat.isDirectory();
      const bothFiles = srcStat.isFile() && destStat.isFile();

      if (srcStat.mode !== destStat.mode) {
        try {
          await access(dest, srcStat.mode);
        } catch (err) {
          // EINVAL access errors sometimes happen which shouldn't because node shouldn't be giving
          // us modes that aren't valid. investigate this, it's generally safe to proceed.
        }
      }

      if (bothFiles && srcStat.size === destStat.size && +srcStat.mtime === +destStat.mtime) {
        // we can safely assume this is the same file
        onDone();
        return;
      }

      if (bothSymlinks && await readlink(src) === await readlink(dest)) {
        // if both symlinks are the same then we can continue on
        onDone();
        return;
      }

      if (bothFolders && !noExtraneous) {
        // mark files that aren't in this folder as possibly extraneous
        const destFiles = await readdir(dest);
        invariant(srcFiles, 'src files not initialised');

        for (const file of destFiles) {
          if (srcFiles.indexOf(file) < 0) {
            const loc = path.join(dest, file);
            possibleExtraneous.add(loc);

            if ((await lstat(loc)).isDirectory()) {
              for (const file of await readdir(loc)) {
                possibleExtraneous.add(path.join(loc, file));
              }
            }
          }
        }
      }
    }

    if (srcStat.isSymbolicLink()) {
      onFresh();
      const linkname = await readlink(src);
      actions.push({
        type: 'symlink',
        dest,
        linkname,
      });
      onDone();
    } else if (srcStat.isDirectory()) {
      await mkdirp(dest);

      const destParts = dest.split(path.sep);
      while (destParts.length) {
        files.add(destParts.join(path.sep));
        destParts.pop();
      }

      // push all files to queue
      invariant(srcFiles, 'src files not initialised');
      let remaining = srcFiles.length;
      if (!remaining) {
        onDone();
      }
      for (const file of srcFiles) {
        queue.push({
          onFresh,
          src: path.join(src, file),
          dest: path.join(dest, file),
          onDone: () => {
            if (--remaining === 0) {
              onDone();
            }
          },
        });
      }
    } else if (srcStat.isFile()) {
      onFresh();
      actions.push({
        type: 'file',
        src,
        dest,
        atime: srcStat.atime,
        mtime: srcStat.mtime,
        mode: srcStat.mode,
      });
      onDone();
    } else {
      throw new Error(`unsure how to copy this: ${src}`);
    }
  }
}

export function copy(src: string, dest: string): Promise<void> {
  return copyBulk([{src, dest}]);
}

export async function copyBulk(
  queue: CopyQueue,
  _events?: {
    onProgress?: ?(dest: string) => void,
    onStart?: ?(num: number) => void,
    possibleExtraneous?: PossibleExtraneous,
    ignoreBasenames?: Array<string>,
  },
): Promise<void> {
  const events: CopyOptions = {
    onStart: (_events && _events.onStart) || noop,
    onProgress: (_events && _events.onProgress) || noop,
    possibleExtraneous: _events ? _events.possibleExtraneous : [],
    ignoreBasenames: (_events && _events.ignoreBasenames) || [],
  };

  const actions: CopyActions = await buildActionsForCopy(queue, events, events.possibleExtraneous);
  events.onStart(actions.length);

  const fileActions = actions.filter((action): boolean => action.type === 'file');
  await promise.queue(fileActions, (data): Promise<void> => new Promise((resolve, reject) => {
    const readStream = fs.createReadStream(data.src);
    const writeStream = fs.createWriteStream(data.dest, {mode: data.mode});

    readStream.on('error', reject);
    writeStream.on('error', reject);

    writeStream.on('open', function() {
      readStream.pipe(writeStream);
    });

    writeStream.once('finish', function() {
      fs.utimes(data.dest, data.atime, data.mtime, function(err) {
        if (err) {
          reject(err);
        } else {
          events.onProgress(data.dest);
          resolve();
        }
      });
    });
  }), 4);

  // we need to copy symlinks last as the could reference files we were copying
  const symlinkActions = actions.filter((action): boolean => action.type === 'symlink');
  await promise.queue(symlinkActions, (data): Promise<void> => {
    return symlink(path.resolve(path.dirname(data.dest), data.linkname), data.dest);
  });
}

function _readFile(loc: string, encoding: string): Promise<any> {
  return new Promise((resolve, reject) => {
    fs.readFile(loc, encoding, function(err, content) {
      if (err) {
        reject(err);
      } else {
        resolve(content);
      }
    });
  });
}

export function readFile(loc: string): Promise<string> {
  return _readFile(loc, 'utf8').then(normalizeOS);
}

export function readFileRaw(loc: string): Promise<Buffer> {
  return _readFile(loc, 'binary');
}

export async function readFileAny(files: Array<string>): Promise<?string> {
  for (const file of files) {
    if (await exists(file)) {
      return readFile(file);
    }
  }
  return null;
}

export async function readJson(loc: string): Promise<Object> {
  const file = await readFile(loc);
  try {
    return map(JSON.parse(stripBOM(file)));
  } catch (err) {
    err.message = `${loc}: ${err.message}`;
    throw err;
  }
}

export async function find(filename: string, dir: string): Promise<string | false> {
  const parts = dir.split(path.sep);

  while (parts.length) {
    const loc = parts.concat(filename).join(path.sep);

    if (await exists(loc)) {
      return loc;
    } else {
      parts.pop();
    }
  }

  return false;
}

export async function symlink(src: string, dest: string): Promise<void> {
  try {
    const stats = await lstat(dest);

    if (stats.isSymbolicLink() && await exists(dest)) {
      const resolved = await realpath(dest);
      if (resolved === src) {
        return;
      }
    }

    await unlink(dest);
  } catch (err) {
    if (err.code !== 'ENOENT') {
      throw err;
    }
  }

  try {
    if (process.platform === 'win32') {
      // use directory junctions if possible on win32, this requires absolute paths
      await fsSymlink(src, dest, 'junction');
    } else {
      // use relative paths otherwise which will be retained if the directory is moved
      const relative = path.relative(path.dirname(dest), src);
      await fsSymlink(relative, dest);
    }
  } catch (err) {
    if (err.code === 'EEXIST') {
      // race condition
      await symlink(src, dest);
    } else {
      throw err;
    }
  }
}

export type WalkFiles = Array<{
  relative: string,
  absolute: string,
  basename: string,
  mtime: number,
}>;

export async function walk(
  dir: string,
  relativeDir?: ?string,
  ignoreBasenames?: Set<string> = new Set(),
): Promise<WalkFiles> {
  let files = [];

  let filenames = await readdir(dir);
  if (ignoreBasenames.size) {
    filenames = filenames.filter((name) => !ignoreBasenames.has(name));
  }

  for (const name of filenames) {
    const relative = relativeDir ? path.join(relativeDir, name) : name;
    const loc = path.join(dir, name);
    const stat = await lstat(loc);

    files.push({
      relative,
      basename: name,
      absolute: loc,
      mtime: +stat.mtime,
    });

    if (stat.isDirectory()) {
      files = files.concat(await walk(loc, relative, ignoreBasenames));
    }
  }

  return files;
}

export function normalizeOS(body: string): string {
  return body.replace(/\r\n/g, '\n');
}
