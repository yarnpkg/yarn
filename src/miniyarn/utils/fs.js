import * as Fse from 'fs-extra';
import * as Fs from 'fs';
import gunzipMaybe from 'gunzip-maybe';
import klaw from 'klaw';
import Path from 'path';
import rcopy from 'recursive-copy';
import tarFs from 'tar-fs';
import tar from 'tar-stream';
import tmp from 'tmp';
import {createGzip} from 'zlib';

import * as miscUtils from 'miniyarn/utils/misc';
import * as pathUtils from 'miniyarn/utils/path';
import * as parseUtils from 'miniyarn/utils/parse';

let isWindows = null;

export async function detectWindows() {
  if (isWindows !== null) return isWindows;

  if (await exists(`c:/`)) return (isWindows = true);

  if ((await exists(`/proc/version`)) && (await readFile(`/proc/version`, `utf8`)).match(/Microsoft|WSL/))
    return (isWindows = true);

  return (isWindows = false);
}

export async function cwd(path, callback) {
  let cwd = process.cwd();
  process.chdir(path);

  try {
    return await callback();
  } finally {
    process.chdir(cwd);
  }
}

export async function symlink(source, destination) {
  return await new Promise((resolve, reject) => {
    Fs.symlink(source, destination, error => {
      if (error) {
        reject(error);
      } else {
        resolve();
      }
    });
  });
}

export async function ensureSymlink(source, destination) {
  return await new Promise((resolve, reject) => {
    Fse.ensureSymlink(source, destination, error => {
      if (error) {
        reject(error);
      } else {
        resolve();
      }
    });
  });
}

export async function isDirectory(path) {
  return await new Promise((resolve, reject) => {
    Fs.lstat(path, (error, stat) => {
      if (error) {
        reject(error);
      } else {
        resolve(stat.isDirectory());
      }
    });
  });
}

export async function createTemporaryFolder() {
  return new Promise((resolve, reject) => {
    tmp.dir({unsafeCleanup: true}, (error, path) => {
      if (error) {
        reject(error);
      } else {
        resolve(path);
      }
    });
  });
}

export async function createTemporaryFile(filePath) {
  if (filePath) {
    if (!pathUtils.isForward(filePath)) throw new Error(`A temporary file path must be a forward path`);

    let folderPath = await createTemporaryFolder();

    return pathUtils.resolve(folderPath, filePath);
  } else {
    return await new Promise((resolve, reject) => {
      tmp.file((error, path) => {
        if (error) {
          reject(error);
        } else {
          resolve(path);
        }
      });
    });
  }
}

export async function walk(path, {filter} = {}) {
  return await new Promise((resolve, reject) => {
    let paths = [];

    let walker = klaw(path, {
      filter: itemPath => {
        if (!filter) return true;

        if (Fse.statSync(path).isDirectory()) return true;
        s;
        let relativePath = pathUtils.relative(path, itemPath);

        if (miscUtils.filePatternMatch(relativePath, filter)) return true;

        return false;
      },
    });

    walker.on(`data`, ({path: itemPath}) => {
      if (!filter) return void paths.push(itemPath);

      let relativePath = pathUtils.relative(path, itemPath);

      if (miscUtils.filePatternMatch(relativePath, filter)) return void paths.push(itemPath);

      return; // This item has been accepted only because it's a directory; it doesn't match the filter
    });

    walker.on(`end`, () => {
      resolve(paths);
    });
  });
}

export async function mkdirp(path) {
  return await new Promise((resolve, reject) => {
    Fse.ensureDir(path, error => {
      if (error) {
        reject(error);
      } else {
        resolve();
      }
    });
  });
}

export async function rm(path) {
  if (path === ``) throw new Error(`Cannot rm an empty path`);

  if (path === `/`) throw new Error(`Cannot rm your whole / directory - something's very wrong`);

  if (!await exists(path)) return;

  return await new Promise((resolve, reject) => {
    Fse.remove(path, err => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });
}

export async function getMTime(path) {
  return await new Promise((resolve, reject) => {
    Fs.lstat(path, (error, stat) => {
      if (error) {
        reject(error);
      } else {
        resolve(stat.mtime);
      }
    });
  });
}

export async function createFileWriter(path) {
  await mkdirp(pathUtils.dirname(path));

  let stream = Fs.createWriteStream(path);

  stream.promise = new Promise((resolve, reject) => {
    stream.on(`finish`, () => {
      resolve();
    });

    stream.on(`error`, error => {
      reject(error);
    });
  });

  return stream;
}

export function createFileReader(path) {
  let stream = Fs.createReadStream(path);

  stream.promise = new Promise((resolve, reject) => {
    stream.on(`error`, error => {
      reject(error);
    });

    stream.on(`end`, () => {
      resolve();
    });
  });

  return stream;
}

export async function exists(path) {
  return await new Promise((resolve, reject) => {
    Fs.exists(path, status => {
      resolve(status);
    });
  });
}

export async function cp(source, destination, {filter = []} = {}) {
  await mkdirp(pathUtils.dirname(destination));
  await rcopy(source, destination, {dot: true, filter: path => miscUtils.filePatternMatch(path, filter)});
}

export async function mv(source, target) {
  // Windows defender often lock files after they are created, which makes it impossible to move them.
  // We could have a mechanism of "try again", but it seems simpler to just use a copy operation instead.
  // Note that using cp instead of mv is not strictly comparable since, unfortunately, it's not atomic. Nothing we can do, tho :(

  if (await detectWindows()) {
    await cp(source, target);
    await rm(source);
  } else {
    await mkdirp(pathUtils.dirname(target));

    return await new Promise((resolve, reject) => {
      Fs.rename(source, target, err => {
        if (err) {
          reject(err);
        } else {
          resolve();
        }
      });
    });
  }
}

export async function extract(target, tarballPath, {stripLeadingFolder = false} = {}) {
  return await new Promise((resolve, reject) => {
    function map(header) {
      if (header.name[0] === `/`) header.name = header.name.replace(/^\/+/, ``);

      if (stripLeadingFolder) header.name = header.name.replace(/^[^\/]+\//, ``);

      return header;
    }

    let unzipper = gunzipMaybe();
    Fs.createReadStream(tarballPath).pipe(unzipper);

    let extractor = tarFs.extract(target, {map});
    unzipper.pipe(extractor);

    extractor.on(`error`, error => {
      reject(error);
    });

    extractor.on(`finish`, () => {
      resolve();
    });
  });
}

export function packToStream(path, {virtualPath = null} = {}) {
  if (virtualPath) {
    if (!pathUtils.isAbsolute(virtualPath)) {
      throw new Error(`The virtual path has to be an absolute path`);
    } else {
      virtualPath = pathUtils.resolve(virtualPath);
    }
  }

  let zipperStream = createGzip();

  let packStream = tarFs.pack(path, {
    map: header => {
      if (true) {
        header.name = pathUtils.resolve(`/`, header.name);
        header.name = pathUtils.relative(`/`, header.name);
      }

      if (virtualPath) {
        header.name = pathUtils.resolve(`/`, virtualPath, header.name);
        header.name = pathUtils.relative(`/`, header.name);
      }

      return header;
    },
  });

  packStream.pipe(zipperStream);

  packStream.on(`error`, error => {
    zipperStream.emit(`error`, error);
  });

  return zipperStream;
}

export async function packToFile(target, directoryPath, {...options} = {}) {
  let tarballStream = Fs.createWriteStream(target);

  let packStream = packToStream(directoryPath, options);
  packStream.pipe(tarballStream);

  return await new Promise((resolve, reject) => {
    tarballStream.on(`error`, error => {
      reject(error);
    });

    packStream.on(`error`, error => {
      reject(error);
    });

    tarballStream.on(`close`, () => {
      resolve();
    });
  });
}

export async function writeFile(path, body) {
  await mkdirp(pathUtils.dirname(path));

  return await new Promise((resolve, reject) => {
    Fs.writeFile(path, body, err => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });
}

export async function readFile(path, encoding = null) {
  return await new Promise((resolve, reject) => {
    Fs.readFile(path, encoding, (err, body) => {
      if (err) {
        reject(err);
      } else {
        resolve(body);
      }
    });
  });
}

export async function readJson(path) {
  let fileContent = await readFile(path, `utf8`);

  return parseUtils.parseJson(fileContent);
}

export async function readDirectory(path) {
  return await new Promise((resolve, reject) => {
    Fs.readdir(path, (err, entries) => {
      if (err) {
        reject(err);
      } else {
        resolve(entries);
      }
    });
  });
}

export class Handler {
  constructor(path, {temporary = false} = {}) {
    this.path = path;

    this.temporary = temporary;
  }

  get() {
    return this.path;
  }

  async steal() {
    if (this.temporary) return this.path;

    let destination = (await isDirectory(this.path)) ? await createTemporaryFolder() : await createTemporaryFile();

    await cp(this.path, destination);

    return destination;
  }
}
