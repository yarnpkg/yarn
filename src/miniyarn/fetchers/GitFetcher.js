import invariant from 'invariant';

import {BaseFetcher} from 'miniyarn/fetchers/BaseFetcher';
import {PackageInfo} from 'miniyarn/models/PackageInfo';
import * as execUtils from 'miniyarn/utils/exec';
import * as fsUtils from 'miniyarn/utils/fs';
import * as miscUtils from 'miniyarn/utils/misc';

export class GitFetcher extends BaseFetcher {
  supports(packageLocator, {env}) {
    if (!packageLocator.name || !packageLocator.reference) return false;

    return miscUtils.isGitReference(packageLocator.reference);
  }

  async fetch(packageLocator, {env}) {
    try {
      return await this.fetchViaArchive(packageLocator, {env});
    } catch (error) {
      return await this.fetchViaClone(packageLocator, {env});
    }
  }

  async fetchViaArchive(packageLocator, {env}) {
    throw new Error(`Unsupported`);
  }

  async fetchViaClone(packageLocator, {env}) {
    invariant(packageLocator.name, `This package locator should have a name`);
    invariant(packageLocator.reference, `This package locator should have a reference`);

    let archivePath = await fsUtils.createTemporaryFile();
    let archiveHandler = new fsUtils.Handler(archivePath, {temporary: true});

    let clonePath = await fsUtils.createTemporaryFolder();

    // Split the git repository url and the reference we'll need to checkout
    let [, gitUrl, tagBranchCommit] = packageLocator.reference.match(/^([^#]*)(?:#(.*))$/);

    // Clone the directory, and remove its history since we don't really care about it
    await execUtils.execFile(`git`, [`clone`, `--depth=1`, `--`, gitUrl, `.`], {cwd: clonePath});
    if (tagBranchCommit) await execUtils.execFile(`git`, [`checkout`, `-b`, tagBranchCommit], {cwd: clonePath});
    await fsUtils.rm(`${clonePath}/.git`);

    // Remove some internal files, just in case they have been accidentally committed (?!)
    await fsUtils.rm(`${clonePath}/${env.ARCHIVE_FILENAME}`);
    await fsUtils.rm(`${clonePath}/${env.ATOMIC_FILENAME}`);
    await fsUtils.rm(`${clonePath}/${env.INFO_FILENAME}`);

    // Extract the package.json file info, and save them in our normalized file format
    let packageInfo = new PackageInfo(JSON.parse(await fsUtils.readFile(`${clonePath}/package.json`, `utf8`))).merge(
      packageLocator,
    );
    await fsUtils.writeFile(`${clonePath}/${env.INFO_FILENAME}`, JSON.stringify(packageInfo.toJSON()));

    // Pack everything into a single archive, and forward it
    await fsUtils.packToFile(archivePath, clonePath);

    return {packageInfo, handler: archiveHandler};
  }
}
