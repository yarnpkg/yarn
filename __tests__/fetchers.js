/* @flow */
/* eslint max-len: 0 */

import TarballFetcher, {LocalTarballFetcher} from '../src/fetchers/tarball-fetcher.js';
import BaseFetcher from '../src/fetchers/base-fetcher.js';
import CopyFetcher from '../src/fetchers/copy-fetcher.js';
import GitFetcher from '../src/fetchers/git-fetcher.js';
import {NoopReporter} from '../src/reporters/index.js';
import Config from '../src/config.js';
import mkdir from './_temp.js';
import * as fs from '../src/util/fs.js';

const path = require('path');

async function createConfig(): Promise<Config> {
  const config = new Config(new NoopReporter());
  await config.init();
  return config;
}

jasmine.DEFAULT_TIMEOUT_INTERVAL = 60000;

test('BaseFetcher.fetch', async () => {
  const dir = await mkdir('base-fetcher');
  const fetcher = new BaseFetcher(dir, {
    type: 'base',
    registry: 'npm',
    reference: '',
  }, await createConfig());
  let error;

  try {
    await fetcher.fetch();
  } catch (e) {
    error = e;
  }
  expect(error && error.message).toBe('Not implemented');
});

test('CopyFetcher.fetch', async () => {
  const a = await mkdir('copy-fetcher-a');
  await fs.writeFile(path.join(a, 'package.json'), '{}');
  await fs.writeFile(path.join(a, 'foo'), 'bar');

  const b = await mkdir('copy-fetcher-b');
  const fetcher = new CopyFetcher(b, {
    type: 'copy',
    reference: a,
    registry: 'npm',
  }, await createConfig());
  await fetcher.fetch();
  const content = await fs.readFile(path.join(b, 'package.json'));
  expect(content).toBe('{}');
  const contentFoo = await fs.readFile(path.join(b, 'foo'));
  expect(contentFoo).toBe('bar');
});

test('GitFetcher.fetch fetchFromLocal not in network or cache', async () => {
  const config = await createConfig();
  const dir = await mkdir('git-fetcher');
  const fetcher = new GitFetcher(dir, {
    type: 'git',
    reference: './offline-mirror/fetch-test',
    hash: '8beb0413a8028ca2d52dbb86c75f42069535591b',
    registry: 'npm',
  }, config);
  try {
    await fetcher.fetch();
  } catch (err) {
    expect(err.message).toBe(config.reporter.lang('tarballNotInNetworkOrCache', './offline-mirror/fetch-test', path.resolve('./offline-mirror/fetch-test')));
  }
});

test('GitFetcher.fetch', async () => {
  const dir = await mkdir('git-fetcher');
  const fetcher = new GitFetcher(dir, {
    type: 'git',
    reference: 'https://github.com/sindresorhus/beeper',
    hash: '8beb0413a8028ca2d52dbb86c75f42069535591b',
    registry: 'npm',
  }, await createConfig());
  await fetcher.fetch();
  const name = (await fs.readJson(path.join(dir, 'package.json'))).name;
  expect(name).toBe('beeper');
});

test('TarballFetcher.fetch', async () => {
  const dir = await mkdir('tarball-fetcher');
  const fetcher = new TarballFetcher(dir, {
    type: 'tarball',
    hash: 'a32262ca1e22a3746b970936d3944b4bfd6cb9e9',
    reference: 'https://github.com/sindresorhus/beeper/archive/master.tar.gz',
    registry: 'npm',
  }, await createConfig());

  await fetcher.fetch();
  const name = (await fs.readJson(path.join(dir, 'package.json'))).name;
  expect(name).toBe('beeper');
});

test('TarballFetcher.fetch throws on invalid hash', async () => {
  const dir = await mkdir('tarball-fetcher');
  const url = 'https://github.com/sindresorhus/beeper/archive/master.tar.gz';
  const fetcher = new TarballFetcher(dir, {
    type: 'tarball',
    hash: 'foo',
    reference: url,
    registry: 'npm',
  }, await createConfig());
  let error;
  try {
    await fetcher.fetch();
  } catch (e) {
    error = e;
  }
  expect(error && error.message).toMatchSnapshot();
});

test('TarballFetcher.fetch supports local ungzipped tarball', async () => {
  const dir = await mkdir('tarball-fetcher');
  const fetcher = new LocalTarballFetcher(dir, {
    type: 'tarball',
    hash: '25c5098052a7bd322c7db80c26852e9209f98d4f',
    reference: path.join(__dirname, 'fixtures', 'fetchers', 'tarball', 'ungzipped.tar'),
    registry: 'npm',
  }, await createConfig());
  await fetcher.fetch();
  const name = (await fs.readJson(path.join(dir, 'package.json'))).name;
  expect(name).toBe('beeper');
});

test('TarballFetcher.fetch properly stores tarball of package in offline mirror', async () => {
  const dir = await mkdir('tarball-fetcher');
  const offlineMirrorDir = await mkdir('offline-mirror');

  const config = await createConfig();
  config.registries.npm.config['yarn-offline-mirror'] = offlineMirrorDir;

  const fetcher = new TarballFetcher(dir, {
    type: 'tarball',
    hash: '6f86cbedd8be4ec987be9aaf33c9684db1b31e7e',
    reference: 'https://registry.npmjs.org/lodash.isempty/-/lodash.isempty-4.4.0.tgz',
    registry: 'npm',
  }, config);

  await fetcher.fetch();
  const exists = await fs.exists(path.join(offlineMirrorDir, 'lodash.isempty-4.4.0.tgz'));
  expect(exists).toBe(true);
});

test('TarballFetcher.fetch properly stores tarball of scoped package in offline mirror', async () => {
  const dir = await mkdir('tarball-fetcher');
  const offlineMirrorDir = await mkdir('offline-mirror');

  const config = await createConfig();
  config.registries.npm.config['yarn-offline-mirror'] = offlineMirrorDir;

  const fetcher = new TarballFetcher(dir, {
    type: 'tarball',
    hash: '6f0ab73cdd7b82d8e81e80838b49e9e4c7fbcc44',
    reference: 'https://registry.npmjs.org/@exponent/configurator/-/configurator-1.0.2.tgz',
    registry: 'npm',
  }, config);

  await fetcher.fetch();
  const exists = await fs.exists(path.join(offlineMirrorDir, '@exponent-configurator-1.0.2.tgz'));
  expect(exists).toBe(true);
});
