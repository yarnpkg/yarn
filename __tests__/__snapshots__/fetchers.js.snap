exports[`test TarballFetcher.fetch throws on invalid hash 1`] = `"https://github.com/sindresorhus/beeper/archive/master.tar.gz: Bad hash. Expected \"foo\" but got \"a32262ca1e22a3746b970936d3944b4bfd6cb9e9\" "`;
